src/dawn/SIR/SIR.proto

```diff
@@ -0,0 +1,627 @@
+/*===------------------------------------------------------------------------------*- proto -*-===*\
+ *                          _
+ *                         | |
+ *                       __| | __ ___      ___ ___
+ *                      / _` |/ _` \ \ /\ / / '_  |
+ *                     | (_| | (_| |\ V  V /| | | |
+ *                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+ *
+ *
+ *  This file is distributed under the MIT License (MIT).
+ *  See LICENSE.txt for details.
+ *
+\*===------------------------------------------------------------------------------------------===*/
+
+// @defgroup sir_proto SIR protobuf
+// @brief This is a SIR description for Google's protocol buffers library
+// <a href="https://developers.google.com/protocol-buffers/">Protobuf</a>.
+
+syntax = "proto3";
+
+package dawn.sir.proto;
+
+// @brief Source information
+//
+// `(-1,-1)` indicates an invalid location.
+// @ingroup sir_proto
+message SourceLocation {
+  int32 Line = 1;   // Source line
+  int32 Column = 2; // Column in the `Line`
+}
+
+// @brief Stencil attributes
+//
+// Stencil attributes can be used to adjust optimization and code generation for individual stencil
+// and stencil_functions.
+// @ingroup sir_proto
+message Attributes {
+  bool no_codegen = 1;        // Don't generate code for this stencil
+  bool merge_stages = 2;      // Merge the Stages of this stencil
+  bool merge_do_methods = 3;  // Merge the Do-Methods of this stencil
+  bool merge_temporaries = 4; // Merge the temporaries of this stencil
+  bool use_k_caches = 5;      // Use K-Caches
+}
+
+// @brief Description of a field argument of a Stencil or StencilFunction
+// @ingroup sir_proto
+message Field {
+  string name = 1;        // Name of the field
+  SourceLocation loc = 2; // Source location
+  bool is_temporary = 3;  // Is the field a temporary?
+}
+
+// @brief Directional argument of a StencilFunction
+//
+// Note that this message is merely a placeholder as the actual arguments is encoded as an AST node.
+// @ingroup sir_proto
+message Direction {
+  string name = 1;        // Name of the directional argument
+  SourceLocation loc = 2; // Source location
```

Comment:
I dont understand yet how we reference this and the Offset, dont we need an unique ID or something in order to dereference it? Or should later the compiler go to the SourceLocation in order to evaluate a particular argument?

Reply:
Thses objects, `Direction`, `Offset` and `Field`, are merely *placeholders* used in the `Stencil` and `StencilFunction` argument lists. The Stencil only needs to know the names of it's input fields thus a simple `name` member suffices. For stencil functions, the actual data is encoded in the AST via the `StencilFunctionArgExpr` node.